tests/unit/test_scopes.py

```diff
@@ -148,9 +154,187 @@ def test_scopebuilder_make_mutable_produces_same_strings():
     assert str(sb.make_mutable("bar")) == sb.bar
 
 
+def test_scopebuilder_make_mutable_can_be_optional():
+    sb = ScopeBuilder(str(uuid.UUID(int=0)), known_scopes="foo")
+    assert str(sb.make_mutable("foo", optional=True)) == "*" + sb.foo
+
+
 def test_flows_scopes_creation():
     assert FlowsScopes.resource_server == "flows.globus.org"
     assert (
         FlowsScopes.run
         == "https://auth.globus.org/scopes/eec9b274-0c81-4334-bdc2-54e90e689b9a/run"
     )
+
+
+def test_scope_parsing_rejects_optional_marker_followed_by_space():
+    Scope.parse("*foo")  # okay
+    with pytest.raises(ScopeParseError):
+        Scope.parse("* foo")  # not okay
+
+
+def test_scope_parsing_allows_empty_string():
+    scopes = Scope.parse("")
+    assert scopes == []
+
+
+@pytest.mark.parametrize(
+    "scopestring",
+    [
+        # ending in '*'
+        "foo*",
+        # '*' followed by '[]'
+        "foo*[bar]",
+        "foo *[bar]",
+        "foo [bar*]",
+        # empty brackets
+        "foo[]",
+        # starting with open bracket
+        "[foo]",
+        # double brackets
+        "foo[[bar]]",
+        # unbalanced brackets
+        "foo[",
+        "foo[bar",
+        "foo[bar]]",
+        # space before brackets
+        "foo [bar]",
+        # missing space before next scope string after ']'
+        "foo[bar]baz",
```

Comment:
I don't see the following possibilities (which might not all be invalid):

* `"foo * "` (space after '*')
* `"foo[ bar]"` (space after '[')
* `" foo"` (leading space)
* `"foo]"` (unmatched closing bracket 1)
* `"foo bar]"` (unmatched closing bracket 2)
* `"foo[bar] baz]"` (unmatched closing bracket 3)

The reason there are multiple unmatched closing brackets examples is because each might exercise different types of "reset" failures.

Note that if plus-separated string parsing is added after reviewing a comment above, these tests will need to also be updated.

Reply:
I just checked, and there are two of these which are valid: `" foo"` (equivalent to `"foo"`) and `"foo[ bar]"`. The others are all definitely invalid, and I'm more than happy to add them as test cases; thanks for enumerating them!
I'll also add those two valid cases to our tests, to document/enforce their validity.

At this point, I'm setup with a copy of Auth's scope parser to play with, so I can check these against the service to ensure we agree on things to accept or reject. So if you have other questionable cases that you think up, we can evaluate them and add as positive or negative tests.

Reply:
I just realized there's one more: `*foo`. I don't think this variant is accounted for yet.

Reply:
I just pushed [6a97871](https://github.com/globus/globus-sdk-python/pull/642/commits/6a97871b24709d4d381a9f838a67ad3858e6c4f3), which adds a few test cases for "can parse" as opposed to these for "can't parse".
It includes `*foo` and gives us a place for any future interesting cases which are valid.